kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/status/NMStatusConverter.java

```diff
@@ -60,81 +61,109 @@ public class NMStatusConverter {
                 accessPointsProperties.getActiveAccessPoint(), accessPointsProperties.getAvailableAccessPoints(),
                 supportedChannelsProperties.getCountryCode(), supportedChannelsProperties.getSupportedChannels());
 
         return builder.build();
     }
 
-    private static void setDeviceStatus(NetworkInterfaceStatusBuilder<?> builder, Properties deviceProperties) {
-        builder.withAutoConnect(deviceProperties.Get(NM_DEVICE_BUS_NAME, "Autoconnect"));
-        builder.withFirmwareVersion(deviceProperties.Get(NM_DEVICE_BUS_NAME, "FirmwareVersion"));
-        builder.withDriver(deviceProperties.Get(NM_DEVICE_BUS_NAME, "Driver"));
-        builder.withDriverVersion(deviceProperties.Get(NM_DEVICE_BUS_NAME, "DriverVersion"));
+    private static void setDeviceStatus(NetworkInterfaceStatusBuilder<?> builder,
+            DevicePropertiesWrapper devicePropertiesWrapper) {
+        builder.withAutoConnect(devicePropertiesWrapper.getDeviceProperties().Get(NM_DEVICE_BUS_NAME, "Autoconnect"));
+        builder.withFirmwareVersion(
+                devicePropertiesWrapper.getDeviceProperties().Get(NM_DEVICE_BUS_NAME, "FirmwareVersion"));
+        builder.withDriver(devicePropertiesWrapper.getDeviceProperties().Get(NM_DEVICE_BUS_NAME, "Driver"));
+        builder.withDriverVersion(
+                devicePropertiesWrapper.getDeviceProperties().Get(NM_DEVICE_BUS_NAME, "DriverVersion"));
 
-        UInt32 mtu = deviceProperties.Get(NM_DEVICE_BUS_NAME, "Mtu");
+        UInt32 mtu = devicePropertiesWrapper.getDeviceProperties().Get(NM_DEVICE_BUS_NAME, "Mtu");
         builder.withMtu(mtu.intValue());
 
-        String hwAddress = deviceProperties.Get(NM_DEVICE_BUS_NAME, "HwAddress");
-        builder.withHardwareAddress(getMacAddressBytes(hwAddress));
+        try {
+            String hwAddress = devicePropertiesWrapper.getDeviceProperties().Get(NM_DEVICE_BUS_NAME, NM_HW_ADDRESS);
+            builder.withHardwareAddress(getMacAddressBytes(hwAddress));
+        } catch (DBusExecutionException e) {
+            logger.debug("NetworkManager version equal or lower then 1.24 detected.");
+            switch (devicePropertiesWrapper.getDeviceType()) {
+            case NM_DEVICE_TYPE_ETHERNET:
+                builder.withHardwareAddress(getMacAddressBytes(
+                        devicePropertiesWrapper.getDeviceProperties().Get(NM_DEVICE_WIRED_BUS_NAME, NM_HW_ADDRESS)));
+                break;
+            case NM_DEVICE_TYPE_WIFI:
+                builder.withHardwareAddress(getMacAddressBytes(
+                        devicePropertiesWrapper.getDeviceProperties().Get(NM_DEVICE_WIRELESS_BUS_NAME, NM_HW_ADDRESS)));
+                break;
+            case NM_DEVICE_TYPE_GENERIC:
+            case NM_DEVICE_TYPE_LOOPBACK:
+            default:
+                logger.debug("Setting HW Address to blank.");
+                break;
+
+            }
+        }
```

Comment:
This is not how I would approach the issue.

1. I would emphasise with the code that we change how we get the hwAddress based on the version.
2. There's not actually a mapping between `NMDeviceType` and how we get the address, but on the builder type
3. I would emphasise with the code that we change how we get the hwAddress based on the builder type
4. You did not check if the `Optional` property is present or not.

Therefore:

```suggestion
String hwAddress = "";
        try {
            hwAddress = devicePropertiesWrapper.getDeviceProperties().Get(NM_DEVICE_BUS_NAME, NM_HW_ADDRESS);
        } catch (DBusExecutionException e) {
            logger.debug("NetworkManager version lower then 1.24 detected.");
        }

        if (hwAddress.isEmpty()) {
            // Fallback for NM version < 1.24
            if (!devicePropertiesWrapper.getDeviceSpecificProperties().isPresent()) {
                throw new IllegalArgumentException("No device specific properties found. Cannot retrieve HwAddress");
            }

            if (builder instanceof EthernetInterfaceStatusBuilder) {
                hwAddress = devicePropertiesWrapper.getDeviceSpecificProperties().get().Get(NM_DEVICE_WIRED_BUS_NAME,
                        NM_HW_ADDRESS);
            } else if (builder instanceof WifiInterfaceStatusBuilder) {
                hwAddress = devicePropertiesWrapper.getDeviceSpecificProperties().get().Get(NM_DEVICE_WIRELESS_BUS_NAME,
                        NM_HW_ADDRESS);
            } else { // Loopback
                hwAddress = "00:00:00:00:00:00";
            }
        }

        builder.withHardwareAddress(getMacAddressBytes(hwAddress));
```

Reply:
My goal here is to make it easier to refactor the code for when we'll have the version parsing (i.e. when we will be able to check `if(NM_VERSION.lowerThan("1.24"))` )